File: Sort Techniques/Merge_Sort.cpp
//Merge Sort
#include<iostream>
using namespace std;

#define size 100

void merge(int arr[],int l,int mid,int r)
{
	int sl = mid - l + 1;
	int sr = r - mid;

	int L[sl];
	int R[sr];

	for(int i=0;i<sl;i++)
		L[i] = arr[l+i];
	
	for(int j=0;j<sr;j++)
		R[j] = arr[mid+j+1];

	int i = 0;
	int j = 0;
	int k = l;

	while(i < sl && j < sr)
	{
		if(L[i] <= R[j])
		{
			arr[k] = L[i];
			k++; i++;
		}

		else{
			arr[k] = R[j];
			k++; j++;
		}
	}

	while(i < sl)
	{
		arr[k] = L[i];
		k++;
		i++;
	}

	while(j < sr)
	{
		arr[k] = R[j];
		k++;
		j++;
	}

}

void mergesort(int arr[],int l,int u)
{
	if(l < u)
	{
		int mid = l+(u - l)/2;
		mergesort(arr,0,mid);
		mergesort(arr,mid+1,u);
		merge(arr,l,mid,u);
	}
}

int main()
{
	int n;
	cout<<"Enter the Length of Array"<<endl;
	cin>>n;
	cout<<"Enter the Array"<<endl;
	int arr[n];
	for(int i=0;i<n;i++)
	{
		cin>>arr[i];
	}	
	
	mergesort(arr,0,n-1);

	cout<<"The Sorted array is : ";
	for(int i=0;i<n;i++)
	cout<<arr[i]<<" ";

	return 0;
}

File: Python/Algorithms/Merge_Sort.py
#Merge Sort Algorithm
# Time Complexity: O(NlogN) running time


print("MERGERSORT")

def merge_sort_1(arr):

    if len(arr) > 1:
        left_arr = arr[:len(arr)//2]
        right_arr = arr[len(arr)//2:]

        #recursion
        merge_sort_1(left_arr)
        merge_sort_1(right_arr)

        #merge
        i = 0
        j = 0
        k = 0

        while i < len(left_arr) and j < len(right_arr):
            if left_arr[i] < right_arr[j]:
                arr[k] = left_arr[i]
                i += 1
            
            else:
                arr[k] = right_arr[j]
                j += 1
            k += 1

        while i < len(left_arr):
            arr[k] = left_arr[i]
            i += 1
            k += 1

        while j < len(right_arr):
            arr[k] = right_arr[j]
            j += 1
            k += 1




if __name__ =='__main__':
    # i = input("Enter the numbers with commas")
    L = [2,6,5,1,7,4,3]
    merge_sort_1(L)

    print(L)